renku/service/views/cache.py

```diff
@@ -0,0 +1,247 @@
+from flask_apispec import marshal_with, use_kwargs
+from marshmallow import EXCLUDE
+from patoolib.util import PatoolError
+
+from renku.core.commands.clone import renku_clone
+from renku.service.config import CACHE_UPLOADS_PATH, \
+    INVALID_PARAMS_ERROR_CODE, SUPPORTED_ARCHIVES
+from renku.service.serializers.cache import FileListResponse, \
+    FileListResponseRPC, FileUploadContext, FileUploadRequest, \
+    FileUploadResponse, FileUploadResponseRPC, ProjectCloneContext, \
+    ProjectCloneRequest, ProjectCloneResponse, ProjectCloneResponseRPC, \
+    ProjectListResponse, ProjectListResponseRPC, extract_file
+from renku.service.utils import make_file_path, make_project_path
+from renku.service.views.decorators import accepts_json, handle_base_except, \
+    handle_git_except, handle_renku_except, handle_validation_except, \
+    header_doc, requires_cache, requires_identity
+
+CACHE_BLUEPRINT_TAG = 'cache'
+cache_blueprint = Blueprint('cache', __name__)
+
+
+@marshal_with(FileListResponseRPC)
+@header_doc(description='List uploaded files.', tags=(CACHE_BLUEPRINT_TAG, ))
+@cache_blueprint.route(
+    '/cache/files-list',
+    methods=['GET'],
+    provide_automatic_options=False,
+)
+@handle_base_except
+@handle_git_except
+@handle_renku_except
+@handle_validation_except
+@requires_cache
+@requires_identity
+def list_uploaded_files_view(user, cache):
+    """List uploaded files ready to be added to projects."""
+    files = [
+        f for f in cache.get_files(user) if make_file_path(user, f).exists()
+    ]
+
+    response = FileListResponseRPC().load({
+        'result': FileListResponse().load({'files': files})
+    })
+    return jsonify(response)
+
+
+@use_kwargs(FileUploadRequest)
+@marshal_with(FileUploadResponseRPC)
+@header_doc(
+    description='Upload file or archive of files.',
+    tags=(CACHE_BLUEPRINT_TAG, ),
+)
+@cache_blueprint.route(
+    '/cache/files-upload',
+    methods=['POST'],
+    provide_automatic_options=False,
+)
+@handle_base_except
+@handle_git_except
+@handle_renku_except
+@handle_validation_except
+@requires_cache
+@requires_identity
+def upload_file_view(user, cache):
+    """Upload file or archive of files."""
+    file = extract_file(request)
+
+    response_builder = {
+        'file_name': file.filename,
+        'content_type': file.content_type,
+        'is_archive': file.content_type in SUPPORTED_ARCHIVES
+    }
+    response_builder.update(FileUploadRequest().load(request.args))
+
+    user_cache_dir = CACHE_UPLOADS_PATH / user['user_id']
+    user_cache_dir.mkdir(exist_ok=True)
+
+    file_path = user_cache_dir / file.filename
+    if file_path.exists():
+        if response_builder.get('override_existing', False):
+            file_path.unlink()
+        else:
+            return jsonify(
+                error={
+                    'code': INVALID_PARAMS_ERROR_CODE,
+                    'reason': 'file exists',
+                }
+            )
+
+    file.save(str(file_path))
```

Comment:
if there's any exception past this point (like extraction failing), the file will be on disk but not in the cache, leaving the disk in a dirty state. Would be cleaner to delete the file if there's a failure further down from here.

Reply:
Right, the other way around is also the same problem so it's not really much cleaner since both solutions would use some deferring mechanism. Reason why I din't opt into complicating this here is that there is a more general solutions to this which (one of) is described in #851.